### ISSUE TITLE
Add attention mask to global attn at training/test. 

### ISSUE THREAD

**Issue by ketranm** [OpenNMT/OpenNMT-py#272]
I think the attention mask over source words isn't used in current version of OpenNMT-py, ie `GlobalAttention.applyMask` is never called. It was there in the previous one.
This makes training incorrect and decoding with batch_size > 1 terrible.

**Comment by srush** [comment]
Are you seeing this in practice? It didn't change my results to remove it.  

Attention mask only at decoding seems hacky. If we add it back, let's do it everywhere.

**Comment by srush** [comment]
For when we implement this, let's do it like it is done with transformer. 

Change RNNDecoderState to save the `src` like  https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/Transformer.py#L284

Compute the src_pad_mask in the RNN decoder like: https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/Transformer.py#L255

Augment GlobalAttention to take a mask like : https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/MultiHeadedAttn.py

This way Translator does not need to know about this. This could be done as part of https://github.com/OpenNMT/OpenNMT-py/issues/268

**Comment by ketranm** [comment]
My result is sensible with batch_size 1. Maybe having explicit use of attention masks is useful for multiseq2multiseq task.

**Comment by srush** [comment]
okay, yeah then I agree this is a bug if you are seeing different results. wonder why mine did not change. We will fix asap.
